Add account verification for users

james-woo opened the pull request.

james-woo committed:
Add welcome page

james-woo committed:
Remove .container

cpjk committed:
Merge pull request #25 from james-woo/accounts

Add welcome page

stvhwrd committed:
Fix spelling

cpjk committed:
Add bookable and timeSlot models

cpjk committed:
Add bookable API routes

stvhwrd committed:
spelling, wording of homepage items

stvhwrd committed:
spelling, wording of welcome message

stvhwrd committed:
Merge branch 'master' of github.com:stvhwrd/average-joes

stvhwrd committed:
merged conflict

cpjk committed:
Merge pull request #28 from stvhwrd/master

minor spelling / wording changes to homepage

cpjk committed:
Add bookableType model

cpjk committed:
Add bookableTypes routes

cpjk committed:
Add bookableType to bookable on creation

cpjk committed:
Add moment.js

cpjk committed:
Merge pull request #32 from cpjk/add_reservations

Add bookables and bookable types

james-woo committed:
Add account verification for users

james-woo committed:
Add account verification for users

james-woo committed:
Add account verification for users

james-woo committed:
Add account verification for users

james-woo commented:
Steps to test:
1. Create a new account with a real email address
2. Try to log in, you should get an error saying you need to validate
3. Go validate the account through the email sent
4. Try and log in again, you should be able to now

cpjk commented on app.js:
```diff
@@ -14,6 +14,7 @@ var sessionApiRoutes = require('./app/routes/api/session');
 var usersApiRoutes = require('./app/routes/api/users');
 var bookableApiRoutes = require('./app/routes/api/bookable');
 var bookableTypeApiRoutes = require('./app/routes/api/bookableType');
+var usersVerify = require('./app/routes/api/usersVerify');
```
Please rename this variable to usersVerifyRoutes for consistency

cpjk commented on app.js:
```diff
@@ -46,6 +47,7 @@ app.use('/api/session', sessionApiRoutes);
 app.use('/api/users', usersApiRoutes);
 app.use('/api/bookables', bookableApiRoutes);
 app.use('/api/bookabletypes', bookableTypeApiRoutes);
+app.use('/verify', usersVerify); 
```
same as above

cpjk replied:
Please make all API routes start with /api/

cpjk commented on app/routes/api/session.js:
```diff
@@ -2,11 +2,40 @@ var express = require('express');
 // Creates a new session for the user if successful
 router.post('/login', passport.authenticate('local'), function(req, res){
-  res.json({currentUser: req.user});
+  if(req.user.confirmed == "false"){
```
So, everytime a non-confirmed user tries to log in it will send another confirmation email? Seems like there should be  a better way.

james-woo replied:
I don't know what other way, I have thought about it a lot. Do you have other suggestions? It was the simplest method I could think of.

james-woo replied:
Actually, I have an idea. Let me try something out before merging.

james-woo replied:
Changes made

cpjk commented on app/routes/api/usersVerify.js:
```diff
@@ -0,0 +1,24 @@
+      res.render('layout');
+    }
+  } else{
+    res.end("<h1>Request is from unknown source</h1>");
```
res.end?

cpjk commented on public/app/controllers/userController.js:
```diff
@@ -20,24 +20,28 @@ userControllers.controller('userShowController', function($http, $routeParams, U
+userControllers.controller('userNewController', function($http, $location, User) {
   var vm = this;
   vm.success = "";
+  var rand = ("0000" + (Math.random()*Math.pow(36,4) << 0).toString(36)).slice(-4);
```
We should use a crypto library function instead of this homebrew stuff.

james-woo committed:
Rename routing variables

Made all API routes start with /api/
Did not change the non-confirmed user log in attempt
Changed res.end to res.redirect
Now using crypto library function and implemented it in backend
instead of front end

james-woo committed:
Rename routing variables

Made all API routes start with /api/
Did not change the non-confirmed user log in attempt
Changed res.end to res.redirect
Now using crypto library function and implemented it in backend
instead of front end

james-woo committed:
Rename routing variables

Made all API routes start with /api/
Did not change the non-confirmed user log in attempt
Changed res.end to res.redirect
Now using crypto library function and implemented it in backend
instead of front end

james-woo committed:
Rename routing variables

Made all API routes start with /api/
Did not change the non-confirmed user log in attempt
Changed res.end to res.redirect
Now using crypto library function and implemented it in backend
instead of front end

james-woo committed:
Change resend email

Changed the way to resend email, the user can now click a link, instead
of resending a link everytime.

james-woo committed:
Change resend email

Changed the way to resend email, the user can now click a link, instead
of resending a link everytime.

james-woo committed:
Change resend email

Changed the way to resend email, the user can now click a link, instead
of resending a link everytime.

james-woo committed:
Change resend email

Changed the way to resend email, the user can now click a link, instead
of resending a link everytime.

james-woo committed:
Change userController function name

Changed confirmAccount to sendConfirmationEmail

james-woo committed:
Change userController function name

Changed confirmAccount to sendConfirmationEmail

james-woo committed:
Change userController function name

Changed confirmAccount to sendConfirmationEmail

james-woo committed:
Add user password reset

User is now able to reset their password
Admin role has also been added.

james-woo committed:
Add user password reset

User is now able to reset their password
Admin role has also been added.

james-woo committed:
Add user password reset

User is now able to reset their password
Admin role has also been added.

james-woo committed:
Add user password reset

User is now able to reset their password
Admin role has also been added.

cpjk commented on public/app/views/partials/session/login.jade:
```diff
@@ -10,6 +10,9 @@ div.text-center
       input(type="password" required placeholder="Password" ng-model="sessionCtrl.password").form-control
     button(type="submit").btn.btn-default.
       Login
+    a(href="http://localhost:3000/forgot/") Forgot?
```
This link will break when we deploy. You should just be able to use "/forgot"

cpjk commented on public/app/views/partials/session/login.jade:
```diff
@@ -10,6 +10,9 @@ div.text-center
 
   div(ng-model="sessionCtrl.message")
-    {{sessionCtrl.message}}
+    a(href="http://localhost:3000/api/verify/resend/{{sessionCtrl.user.key}}", target='_blank')
```
same as above

cpjk commented on app/routes/api/session.js:
```diff
@@ -15,6 +29,50 @@ router.post('/logout', function(req, res){
+    if(err){
+      console.log(err);
+    }
+    if(user == null) {
```
You can just use if(!user) here

cpjk commented on app/routes/api/users.js:
```diff
@@ -33,18 +85,38 @@ router.get('/:username', function(req, res, next){
-  User.register(new User({username: req.body.username, firstname: req.body.firstname, lastname: req.body.lastname, email: req.body.email}), req.body.password, function(err, user){
-
+  var hash = crypto.randomBytes(20).toString('hex');
+  User.register(new User({username: req.body.username, firstname: req.body.firstname, lastname: req.body.lastname, email: req.body.email, key: hash, confirmed: "false", permissions: "user"}), req.body.password, function(err, user){
```
Please reformat this so it is like 
{
username: asdasd,
stuff: asdasd,
asdasd: asdasd
}

cpjk replied:
Please reformat this so it is like 
{
username: asdasd,
stuff: asdasd,
asdasd: asdasd
}

cpjk commented on app/routes/api/users.js:
```diff
@@ -33,18 +85,38 @@ router.get('/:username', function(req, res, next){
       passport.authenticate('local')(req, res, function(){
-        res.json({success: true});
+      res.status(200).json({success: true, key: hash})
+      //res.json({success: true});
```
Remove comment.

james-woo committed:
Change routing

james-woo committed:
Change routing

james-woo committed:
Change routing

cpjk committed:
Merge pull request #23 from ProgrammingJoe/profile_page

Profile page and Edit page

cpjk commented on app/routes/api/usersPasswordReset.js:
```diff
@@ -0,0 +1,62 @@
+});
+
+router.get('/:key',function(req,res){
+  var password = crypto.randomBytes(10).toString('hex');
```
Is there a reason this one is 10  bytes and the other is 20?

cpjk commented on app/models/user.js:
```diff
@@ -1,12 +1,16 @@
 var mongoose = require('../../app/models/mongoose');
 var passportLocalMongoose = require('passport-local-mongoose');
+var crypto = require('crypto');
```
Doesn't look like this require is used

james-woo committed:
Change user checking

james-woo committed:
Change user checking

james-woo committed:
Merge branch 'email_verification' of https://github.com/james-woo/average-joes into email_verification

Conflicts:
	app/routes/api/session.js

cpjk commented on app/models/bookable.js:
```diff
@@ -0,0 +1,9 @@
+var mongoose = require('../../app/models/mongoose');
```
this file should not be in this PR

cpjk commented on app/models/bookableType.js:
```diff
@@ -0,0 +1,7 @@
+var mongoose = require('../../app/models/mongoose');
```
same

cpjk commented on app/models/timeSlot.js:
```diff
@@ -0,0 +1,8 @@
+var mongoose = require('../../app/models/mongoose');
```
same

james-woo committed:
Merge branch 'email_verification' of https://github.com/james-woo/average-joes into email_verification

Conflicts:
	app.js
	app/routes/api/users.js
	public/app/controllers/userController.js
	public/app/views/partials/session/login.jade
